mesonbuild/dependencies/base.py

```diff
@@ -244,11 +244,20 @@ def __init__(self, name, environment, kwargs):
                     m = 'Invalid version of dependency, need {!r} {!r} found {!r}.'
                     raise DependencyException(m.format(name, not_found, self.version))
                 return
-        found_msg += [mlog.green('YES'), self.version]
-        # Fetch cargs to be used while using this dependency
-        self._set_cargs()
-        # Fetch the libraries and library paths needed for using this
-        self._set_libs()
+
+        try:
+            # Fetch cargs to be used while using this dependency
+            self._set_cargs()
+            # Fetch the libraries and library paths needed for using this
+            self._set_libs()
+            found_msg += [mlog.green('YES'), self.version]
+        except DependencyException as e:
+            if self.required:
+                raise
+            else:
+                self.is_found = False
```

Comment:
Set `self.compile_args` and `self.link_args` to `[]` here to be sure no bad state escapes by accident. Thanks.